File: C++ Implementation/pw_generator.cpp
/*-----------------------------------------------------------------------------
credit to http://www.zedwood.com/article/cpp-sha256-function for sha256
implementation 

simple script to generate secure password from keywords. Useful for those
who want a password that's not susceptible to dictionary attack, don't trust
password managers, and/or prefer to memorize several short strings instead
of a password

C++ implementation
Usage: ./pw_generator hello entire world
-----------------------------------------------------------------------------*/

#include <iostream>
#include <algorithm>
#include <bits/stdc++.h>
#include "sha256.h"

using namespace std;

int main(int argc, char *argv[])
{
  //user should change these
  const int pw_len = 16;
  const string SALT = "cumin";
  const char SPEC_CHAR_1 = '!';
  const char SPEC_CHAR_2 = '@';
  const int CHAR_POS = 7;
  assert(CHAR_POS < pw_len/2);

  if (argc == 1)
  {
    cout << "Usage: ./pw_generator [keyword1] [keyword2] [keyword3]..." <<endl;
  }
  else if (argc < 4)
  {
    cout << "Minimum 2 keywords" << endl;
  }
  else
  {
    cout << "argc: " << argc << endl;

    //initialize array of args
    string input [argc];
    for (int i=1; i < argc; i++)
    {
      input[i - 1] = argv[i];
    }
    input[argc] = SALT; //add salt
    string sorted_arg_str; 
    sort(input, input+(argc - 1));

    //joins sorted arg array into string
    for (int i=0; i <= argc; i++)
    {
      sorted_arg_str += input[i];
    }

    //hash sorted arg, take substring of set length
    string hash_obj = sha256(sorted_arg_str);
    string shortened_hash = hash_obj.substr(0, pw_len);

    shortened_hash[CHAR_POS] = SPEC_CHAR_1;
    shortened_hash[CHAR_POS * 2] = SPEC_CHAR_2; 

    string front_hash = shortened_hash.substr(0, pw_len/2);
    string back_hash = shortened_hash.substr(pw_len/2, pw_len);

    //capitalizes first half of shortened hash ojb
    transform(front_hash.begin(), front_hash.end(), front_hash.begin(), 
      ::toupper);

    string final_hash = front_hash + back_hash;
    cout << final_hash << endl;

  }

  return 0;

}

File: pw_generator.py
#!/usr/bin/python3

#------------------------------------------------------------------------------
# simple script to generate secure password from keywords. Useful for those
# who want a password that's not susceptible to dictionary attack, don't trust
# password managers, and/or prefer to memorize several short strings instead
# of a password
#
# python implementation
# Usage: ./pw_generator.py hello entire world
#------------------------------------------------------------------------------

import sys
import hashlib

# settings for each user to better randomize & personalize hash results
PW_len = 16  # recommend 10-20
SALT = 'cumin'  # random word to include in options to randomize hash results
SPEC_CHAR_1 = '!'  # !@#$%^&*()+ are good options
SPEC_CHAR_2 = '@'
char_pos = 7
assert char_pos in range(1, PW_len//2)

if len(sys.argv) == 1:
    print('Usage: ./pw_generator.py [keyword1] [keyword2] [keyword3] ...')
elif len(sys.argv) <= 3:
    print('Minimum 2 keywords')  # just good practice
else:
    args = sys.argv[1:]
    args.sort()
    joined_args = ''.join(args)
    # hash function only takes in byte obj
    byte_args = str.encode(joined_args + SALT)
    hash_obj = hashlib.sha256(byte_args)  # see hashlib docs for alternatives
    hash_res = hash_obj.hexdigest()[:16]

    # subs in special char
    hash_res = hash_res.replace(hash_res[char_pos], SPEC_CHAR_1)
    hash_res = hash_res.replace(hash_res[char_pos*2], SPEC_CHAR_2)

    # capitalizes portion of hash - halves hash, capitalizes first half, rejoin
    h1, h2 = hash_res[:len(hash_res)//2], hash_res[len(hash_res)//2:]
    h1 = h1.upper()
    final_hash = str(h1+h2)
    print(final_hash)
